File: JavaScript/Parte1.js
const puppeteer = require('puppeteer');
const ObjectsToCsv = require('objects-to-csv');

async function scrapeGoogleScholarCitations(url) {
  try {
    // Lanzar Puppeteer con la nueva implementación headless
    const browser = await puppeteer.launch({ headless: "new" });
    const page = await browser.newPage();
    await page.goto(url);

    const articles = [];

    async function scrapeArticleInfo() {
      // Selector para los enlaces a los artículos
      const articlesOnPage = await page.$$('h3 a');

      for (const articleLink of articlesOnPage) {
        const article = {};
        article.Title = await articleLink.evaluate(node => node.innerText);

        // Navegar al enlace del artículo
        const articleUrl = await articleLink.getProperty('href');
        await page.goto(articleUrl);

        // Capturar información del artículo
        await page.waitForSelector('a[href*="/scholar.bib"]');
        const bibLink = await page.$('a[href*="/scholar.bib"]');
        article.BibTeX = await bibLink.evaluate(node => node.getAttribute('href'));

        const infoSections = await page.$$('div.gs_a');
        const infoText = await infoSections[0].evaluate(node => node.innerText);
        const infoParts = infoText.split('-');
        article.Author = infoParts[0].trim();
        article.Year = infoText.match(/\d{4}/)[0];
        article.Journal = infoParts[1].trim();
        article.Publisher = (await infoSections[1].evaluate(node => node.innerText)).replace(' -', '').trim();

        article.URL = articleUrl;

        const citations = await page.$('a[href*="cites="]');
        article['Citado por'] = (await citations.evaluate(node => node.innerText)).replace('Citado por ', '');

        articles.push(article);

        // Regresar a la página anterior
        await page.goBack();
      }
    }

    // Realizar el scraping de la página actual
    await scrapeArticleInfo();

    // Navegar a las siguientes páginas
    while (true) {
      const nextPageButton = await page.$('span.gs_ico + a');
      if (nextPageButton) {
        await nextPageButton.click();
        await page.waitForTimeout(2000); // Esperar a que cargue la página
        await scrapeArticleInfo();
      } else {
        break;
      }
    }

    await browser.close();

    // Guardar los datos en un archivo CSV
    const csv = new ObjectsToCsv(articles);
    await csv.toDisk('web_scraping_citas.csv');

    console.log('Web scraping completado y datos guardados en web_scraping_citas.csv');
  } catch (error) {
    console.error('Error durante el scraping:', error);
  }
}

const url = 'URL de Google Scholar aquí'; // Reemplaza con la URL que desees
scrapeGoogleScholarCitations('https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=cancer&btnG=');

File: R/Parte1.r
# Cargar las bibliotecas necesarias
library(rvest)
library(dplyr)

# Definir la URL de Google Scholar
url <- "https://scholar.google.com/scholar?cites=5866269323493626547&as_sdt=2005&sciodt=0,5&hl=es"

# Realizar la función para extraer información
extract_info <- function(url) {
  page <- read_html(url)
  
  # Extraer información de la página actual
  articles <- page %>%
    html_nodes(".gs_ri") %>%
    lapply(function(article) {
      title <- article %>%
        html_node("h3 a") %>%
        html_text()
      author <- article %>%
        html_node(".gs_a") %>%
        html_text()
      journal <- article %>%
        html_node(".gs_a") %>%
        html_text() %>%
        str_match(".*?-(.*?)-.*?\\d{4}") %>%
        .[,2]
      year <- article %>%
        html_node(".gs_a") %>%
        html_text() %>%
        str_match("(\\d{4})") %>%
        .[,2]
      publisher <- article %>%
        html_node(".gs_a") %>%
        html_text() %>%
        str_match("-(.*?)-") %>%
        .[,2]
      url <- article %>%
        html_node("h3 a") %>%
        html_attr("href")
      cited_by <- article %>%
        html_node(".gs_fl a") %>%
        html_text() %>%
        str_match("\\d+") %>%
        .[,1]
      
      data.frame(
        Title = title,
        Author = author,
        Journal = journal,
        Year = year,
        Publisher = publisher,
        URL = url,
        Cited_By = cited_by
      )
    })
  
  return(articles)
}

# Realizar el web scraping y almacenar los datos en un DataFrame
data <- data.frame()
current_url <- url

while (TRUE) {
  articles <- extract_info(current_url)
  data <- rbind(data, do.call(rbind, articles))
  
  next_page <- page %>%
    html_node(".gs_ico_nav_next a") %>%
    html_attr("href")
  
    # Agregar un retraso de 5 segundos
    Sys.sleep(5)

  if (is.na(next_page)) {
    break
  }
  
  current_url <- paste0("https://scholar.google.com", next_page)
}

# Almacenar los datos en un archivo CSV
write.csv(data, "web_scraping_citas.csv", row.names = FALSE)